clientv3/retry.go

```diff
@@ -23,65 +23,96 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/transport"
 )
 
 type rpcFunc func(ctx context.Context) error
 type retryRpcFunc func(context.Context, rpcFunc) error
-type retryStopErrFunc func(error) bool
 
-func isReadStopError(err error) bool {
-	eErr := rpctypes.Error(err)
-	// always stop retry on etcd errors
-	if _, ok := eErr.(rpctypes.EtcdError); ok {
-		return true
-	}
-	// only retry if unavailable
-	ev, _ := status.FromError(err)
-	return ev.Code() != codes.Unavailable
-}
-
-func isWriteStopError(err error) bool {
-	ev, _ := status.FromError(err)
-	if ev.Code() != codes.Unavailable {
-		return true
-	}
-	return rpctypes.ErrorDesc(err) != "there is no address available"
-}
-
-func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRpcFunc {
+func (c *Client) newRetryWrapper(isWrite bool) retryRpcFunc {
 	return func(rpcCtx context.Context, f rpcFunc) error {
 		for {
-			select {
-			case <-c.balancer.ConnectNotify():
-			case <-rpcCtx.Done():
-				return rpcCtx.Err()
-			case <-c.ctx.Done():
-				return c.ctx.Err()
-			}
 			pinned := c.balancer.pinned()
 			err := f(rpcCtx)
 			if err == nil {
 				return nil
 			}
 			if logger.V(4) {
-				logger.Infof("clientv3/retry: error %v on pinned endpoint %s", err, pinned)
+				logger.Infof("clientv3/retry: error %v on pinned endpoint %s (write %v)", err, pinned, isWrite)
+			}
+
+			// error from etcd server with codes.Unavailable, then endpoint switch and retry
+			// e.g. rpctypes.ErrTimeout
+			// error from etcd server with non codes.Unavailable, then no endpoint switch and no retry
+			// e.g. rpctypes.ErrEmptyKey, rpctypes.ErrNoSpace
+			rerr := rpctypes.Error(err)
+			if ev, ok := rerr.(rpctypes.EtcdError); ok {
+				if ev.Code() != codes.Unavailable {
+					return err
+				}
+				if c.balancer.endpointError(pinned, err) {
+					c.balancer.next()
+				}
+				// only retry immutable RPCs ("put at most once semantics")
+				if !isWrite {
+					select {
+					case <-c.balancer.ConnectNotify():
+					case <-rpcCtx.Done():
+						return rpcCtx.Err()
+					case <-c.ctx.Done():
+						return c.ctx.Err()
+					}
+					continue
+				}
+				return err
 			}
-			// mark this before endpoint switch is triggered
-			c.balancer.endpointError(pinned, err)
-			notify := c.balancer.ConnectNotify()
-			if s, ok := status.FromError(err); ok && s.Code() == codes.Unavailable {
-				c.balancer.next()
+
+			// if temporary gRPC connection error, then endpoint switch but no retry
+			// e.g. transport.ErrConnClosing when server closed transport, failing node
+			if tv, ok := err.(transport.ConnectionError); ok && tv.Temporary() {
```

Comment:
if it is temporary, we should retry on the same endpoint not switch to another one.

Reply:
if it is not, we should retry on another endpoint if possible. also it is not clear to me if gRPC will ever return connection error. can you verify it?